app/controllers/api/account.php

```diff
@@ -1647,6 +1654,99 @@
         throw new Exception('Session not found', 404);
     });
 
+App::patch('/v1/account/sessions/:sessionId/oauth2-tokens')
+    ->desc('Update OAUth2 Tokens')
+    ->groups(['api', 'account'])
+    ->label('scope', 'account')
+    ->label('event', 'account.sessions.update')
+    ->label('sdk.auth', [APP_AUTH_TYPE_SESSION, APP_AUTH_TYPE_JWT])
+    ->label('sdk.namespace', 'account')
+    ->label('sdk.method', 'updateOAuth2Tokens')
+    ->label('sdk.description', '/docs/references/account/update-oauth2-tokens.md')
+    ->label('sdk.response.code', Response::STATUS_CODE_OK)
+    ->label('sdk.response.type', Response::CONTENT_TYPE_JSON)
+    ->label('sdk.response.model', Response::MODEL_SESSION)
+    ->label('abuse-limit', 10)
+    ->param('sessionId', null, new UID(), 'Session ID. Use the string \'current\' to update the current device session.')
+    ->param('force', false, new Boolean(), 'Should generate new token even if current one is still valid?', true)
+    ->inject('request')
+    ->inject('response')
+    ->inject('user')
+    ->inject('dbForProject')
+    ->inject('project')
+    ->inject('locale')
+    ->inject('audits')
+    ->inject('events')
+    ->inject('usage')
+    ->action(function ($sessionId, $force, $request, $response, $user, $dbForProject, $project, $locale, $audits, $events, $usage) {
+        /** @var Appwrite\Utopia\Request $request */
+        /** @var boolean $force */
+        /** @var Appwrite\Utopia\Response $response */
+        /** @var Utopia\Database\Document $user */
+        /** @var Utopia\Database\Database $dbForProject */
+        /** @var Utopia\Database\Document $project */
+        /** @var Utopia\Locale\Locale $locale */
+        /** @var Appwrite\Event\Event $audits */
+        /** @var Appwrite\Event\Event $events */
+        /** @var Appwrite\Stats\Stats $usage */
+
+        $sessionId = ($sessionId === 'current')
+            ? Auth::sessionVerify($user->getAttribute('sessions'), Auth::$secret)
+            : $sessionId;
+
+        $sessions = $user->getAttribute('sessions', []);
+
+        foreach ($sessions as $key => $session) {/** @var Document $session */
+            if ($sessionId == $session->getId()) {
+                $expireAt = (int) $session->getAttribute('providerAccessTokenExpiry');
+                if(\time() < $expireAt - 5 && !$force) { // 5 seconds time-sync and networking gap, to be safe
+                    return $response->noContent();
+                }
+
+                $provider = $session->getAttribute('provider');
+                $refreshToken = $session->getAttribute('providerRefreshToken');
+
+                $appId = $project->getAttribute('providers', [])[$provider.'Appid'] ?? '';
+                $appSecret = $project->getAttribute('providers', [])[$provider.'Secret'] ?? '{}';
+
+                $className = 'Appwrite\\Auth\\OAuth2\\'.\ucfirst($provider);
+                $oauth2 = new $className($appId, $appSecret, '', [], []);
```

Comment:
We're missing a check to see if provider is actually supported.

```php
        $className = 'Appwrite\\Auth\\OAuth2\\' . \ucfirst($provider);
        if (!\class_exists($className)) {
            throw new Exception('Provider is not supported', 501);
        }
```

Reply:
Great catch, thanks.